dupedetection/service.go

```diff
@@ -0,0 +1,279 @@
+	insertFingerprintStatement    = `INSERT INTO image_hash_to_image_fingerprint_table(sha256_hash_of_art_image_file, model_1_image_fingerprint_vector, model_2_image_fingerprint_vector, model_3_image_fingerprint_vector, model_4_image_fingerprint_vector) VALUES("%s", x'%s', x'%s', x'%s', x'%s')`
+)
+
+type dupeDetectionFingerprints struct {
+	Sha256HashOfArtImageFile           string    `json:"sha256_hash_of_art_image_file,omitempty"`
+	PathToArtImageFile                 string    `json:"path_to_art_image_file,omitempty"`
+	NumberOfBlock                      int       `json:"number_of_block,omitempty"`
+	DatetimeFingerprintAddedToDatabase time.Time `json:"datetime_fingerprint_added_to_database,omitempty"`
+	Model1ImageFingerprintVector       []float64 `json:"model_1_image_fingerprint_vector,omitempty"`
+	Model2ImageFingerprintVector       []float64 `json:"model_2_image_fingerprint_vector,omitempty"`
+	Model3ImageFingerprintVector       []float64 `json:"model_3_image_fingerprint_vector,omitempty"`
+	Model4ImageFingerprintVector       []float64 `json:"model_4_image_fingerprint_vector,omitempty"`
+}
+
+type service struct {
+	config       *Config
+	pastelClient pastel.Client
+	p2pClient    p2p.Client
+	db           *db.DB
+}
+
+// Run starts task
+func (s *service) Run(ctx context.Context) error {
+	ctx = log.ContextWithPrefix(ctx, "dd-service")
+	defer s.db.Close()
+
+	if err := s.waitSynchronization(ctx); err != nil {
+		return errors.Errorf("failed to wait synchronization, err: %w", err)
+	}
+
+	if err := s.runTask(ctx); err != nil {
+		log.WithContext(ctx).Errorf("failed to run task, err: %s", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Errorf("context done: %w", ctx.Err())
+		case <-time.After(runTaskIntervalMin * time.Minute):
+			if err := s.runTask(ctx); err != nil {
+				log.WithContext(ctx).Errorf("failed to run task, err: %s", err)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *service) waitSynchronization(ctx context.Context) error {
+	checkTimeout := func(checked chan<- struct{}) {
+		time.Sleep(synchronizationTimeoutSec * time.Second)
+		close(checked)
+	}
+
+	timeoutCh := make(chan struct{})
+	go checkTimeout(timeoutCh)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Errorf("context done: %w", ctx.Err())
+		case <-time.After(synchronizationIntervalSec * time.Second):
+			st, err := s.pastelClient.MasterNodeStatus(ctx)
+			if err != nil {
+				log.WithContext(ctx).Errorf("failed to get status from master node, err: %s", err)
+				return err
+			}
+
+			if st.Status == masterNodeSuccessfulStatus {
+				log.WithContext(ctx).Debug("done for waiting synchronization status")
+				return nil
+			}
+		case <-timeoutCh:
+			return errors.Errorf("timeout")
+		}
+	}
+}
+
+func (s *service) getLatestFingerprint(ctx context.Context) (*dupeDetectionFingerprints, error) {
+	row, err := s.db.QueryStringStmt(getLatestFingerprintStatement)
```

Comment:
in case of database is too big, query all items seem not well?
anw, be fine now, but should improve it later

Reply:
noted